lib/semantics/type.cc

```diff
@@ -106,6 +106,30 @@ std::ostream &operator<<(std::ostream &o, const ShapeSpec &x) {
   return o;
 }
 
+bool ArraySpec::IsExplicitShape() const {
+  return CheckAll([](const ShapeSpec &x) { return x.ubound().isExplicit(); });
+}
+bool ArraySpec::IsAssumedShape() const {
+  return CheckAll([](const ShapeSpec &x) { return x.ubound().isDeferred(); });
+}
+bool ArraySpec::IsDeferredShape() const {
+  return CheckAll([](const ShapeSpec &x) {
+    return x.lbound().isDeferred() && x.ubound().isDeferred();
+  });
+}
+bool ArraySpec::IsImpliedShape() const {
+  return CheckAll([](const ShapeSpec &x) { return x.ubound().isAssumed(); });
+}
+bool ArraySpec::IsAssumedSize() const {
+  return !empty() &&
+      std::all_of(begin(), end() - 1,
```

Comment:
`end() - 1` is the same as `back()`, at least when the list isn't empty.  You probably want `end() - 2` here and a length check, or a different implementation.

Reply:
`std::all_of(first, last, predicate)` tests the predicate on `[first, last)`. So this is checking the predicate on all but `back()` and then checking the upper bound of `back()` is assumed. (And `back()` is known to exist because of `!empty()`.)